lib/rex/stopwatch.rb

```diff
@@ -0,0 +1,18 @@
+module Rex
+
+  # This provides a correct way to time an operation provided within a block.
+  #
+  # @see https://blog.dnsimple.com/2018/03/elapsed-time-with-ruby-the-right-way/
+  #
+  # @yield [] The block whose operation should be timed.
+  #
+  # @return Returns the result of the block and the elapsed time in seconds.
+  def self.stopwatch
```

Comment:
Is it worth adding this to a real stopwatch namespace, so we can put additional time recording metrics here in the future?

For context, we've got different scenarios in framework where we want to use an `ensure` to always log out something https://github.com/rapid7/metasploit-framework/blob/f72db79449ffd2d116fb5bce422eedb97d2d9627/lib/msf/ui/console/command_dispatcher/developer.rb#L315-L326 - and it might be good to lump similar methods together

In other places we've wanted to log the times to csv too, so that sort of method might fit nicely in a namespace too

Reply:
Or converting this into a module and using `module_function` would work to IMO 🤔

Reply:
Taking inspiration from csharp's API https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.stopwatch?view=net-5.0, potentially:

```
module Rex::Stopwatch
  def elapsed_milliseconds
    start = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    ret = yield
    elapsed = Process.clock_gettime(Process::CLOCK_MONOTONIC) - start

    [ret, elapsed]
  end
  module_function :elapsed_milliseconds
end
```

Usage:
```
puts Rex::Stopwatch.elapsed_milliseconds { sleep 1 }
```

Again, not opinionated on it being a `module_function` happy to nuke that and stick with `def self.elapsed_milliseconds`

Reply:
I don't have a strong preference. It does make sense we may want to expand on it in the future but I'm not sure what that may look like.

How does a Stopwatch module with a `time_it` method sound?

Reply:
One last API idea; we could potentially keep with the idea of using an argument to influence the base unit:

```
module Rex::Stopwatch
  def self.elapsed_time(unit: :second)
    start = Process.clock_gettime(Process::CLOCK_MONOTONIC, unit)
    ret = yield
    elapsed = Process.clock_gettime(Process::CLOCK_MONOTONIC, unit) - start

    [ret, elapsed]
  end
end
```

So folk can pick and choose what precision they want

```
puts Rex::Stopwatch.elapsed_time(unit: :float_second) { sleep 1 }
puts Rex::Stopwatch.elapsed_time(unit: :float_millisecond) { sleep 1 }
```